### PR TITLE
Use full chain of member expressions when generating display names.

### PR DESCRIPTION
**Pull request by mihaip**
Assumed to be namespaced names (with the exception of "exports", which is special-cased)

### CODE REVIEW

**Comment by syranide** [comment]
The current behavior is how JavaScript generally implies display names though IIRC? You're always free to provide your own displayName if you feel the one generated by JSX is insufficient.

**Comment by mihaip** [comment]
@syranide I don't think JS ever generates display names for property assignments, `foo.bar = function() {...};` will end up with an anonymous function. Also, my understanding is that `displayName` is meant for debugging purposes only, so it does not need to follow JS semantics.

**Comment by spicyj** [comment]
What do you think about doing something like this?

```js
/**
 * If `expr` is an Identifier or MemberExpression node made of identifiers and
 * dot accesses, return a list of the identifier parts. Other nodes return null.
 *
 * Examples:
 *
 * MyComponent -> ['MyComponent']
 * namespace.MyComponent -> ['namespace', 'MyComponent']
 * namespace['foo'] -> null
 * namespace['foo'].bar -> ['bar']
 */
function flattenIdentifierOrMemberExpression(expr) {
  if (expr.type === Syntax.Identifier) {
    return [expr.name];
  } else if (expr.type === Syntax.MemberExpression) {
    if (!expr.computed && expr.property.type === Syntax.Identifier) {
      var flattenedObject = flattenIdentifierOrMemberExpression(expr.object);
      if (flattenedObject) {
        flattenedObject.push(expr.property.name);
        return flattenedObject;
      } else {
        return [expr.property.name];
      }
    }
  }
  return null;
}

...

var displayNamePath = flattenIdentifierOrMemberExpression(left);
if (displayNamePath) {
  if (displayNamePath.length > 1 && displayNamePath[0] === 'exports') {
    displayNamePath.shift();
  }
  addDisplayName(displayNamePath.join('.'), right, state);
}
```

Not tested, but it feels little cleaner to me.

Also, can you add a test for the `exports` behavior too?

File: vendor/fbtransform/transforms/reactDisplayName.js [hunk 1]
```diff
@@ -72,11 +73,40 @@ function visitReactDisplayName(traverse, object, path, state) {
...
+        left.object.name === 'exports') {
+      // Ignore the "exports" property access.
+      if (left.property.type === Syntax.Identifier) {
+        displayName = left.property.name;
+      }
+    } else {
+      var memberExpression = left;
+      var displayNamePath = [];
+      while (memberExpression) {
+        if (memberExpression.property.type === Syntax.Identifier) {
```
Comment (spicyj) [inline vendor/fbtransform/transforms/reactDisplayName.js]: (You need to check `memberExpression.computed` here or else you'll catch `foo[bar]`.)

**Comment by mihaip** [comment]
Updated with your suggestion (worked out of the box).

There was already a test for the `exports` behavior, that did not need to be added.

I also changed the ordering a bit, we were previously computing the candidate display name before we were actually checking that this was a `React.createClass` call. Now that display name computation is more expensive, I've re-ordered such that we only do it if we actually need to.

**Comment by spicyj** [comment]
Thank you!

**Pull request merged**